## Lab01: Thu thập dữ liệu

- Họ tên: Lê Kiệt
- MSSV: 19120554

***
## Cách làm bài và nộp bài

&#9889; Bạn lưu ý là mình sẽ dùng chương trình hỗ trợ chấm bài nên bạn cần phải tuân thủ chính xác qui định mà mình đặt ra, nếu không rõ thì hỏi, chứ không nên tự tiện làm theo ý của cá nhân.

**Cách làm bài**

Bạn sẽ làm trực tiếp trên file notebook này. Đầu tiên, bạn điền họ tên và MSSV vào phần đầu file ở bên trên. Trong file, bạn làm bài ở những chỗ có ghi là:

```python
# YOUR CODE HERE
raise NotImplementedError()
```

Khi làm thì bạn xóa dòng `raise NotImplementedError()` đi.
Đối những phần yêu cầu code thì thường ở ngay phía dưới sẽ có một (hoặc một số) cell chứa các bộ test để giúp bạn biết đã code đúng hay chưa. Nếu chạy cell này không có lỗi gì thì có nghĩa là qua được các bộ test. 

Trong một số trường hợp, các bộ test có thể sẽ không đầy đủ, nghĩa là, nếu không qua được test thì là code sai, nhưng nếu qua được test thì chưa chắc đã đúng.

Trong khi làm bài, bạn có thể cho in ra màn hình, tạo thêm các cell để test. Nhưng khi nộp bài thì bạn xóa các cell mà bạn tự tạo, xóa hoặc comment các câu lệnh in ra màn hình. Bạn lưu ý <font color=red>không được tự tiện xóa các cell hay sửa code của mình </font> (trừ những chỗ được phép sửa như đã nói ở trên).

Trong khi làm bài, thường xuyên `Ctrl + S` để lưu lại bài làm của bạn, tránh mất mát thông tin.

*Nên nhớ mục tiêu chính ở đây là <font color=green>học, học một cách chân thật</font>. Bạn có thể thảo luận ý tưởng với bạn khác cũng như tham khảo các nguồn trên mạng, nhưng sau cùng <font color=green>code và bài làm phải là của bạn, dựa trên sự hiểu thật sự của bạn</font>. Khi tham khảo các nguồn trên mạng thì bạn cần ghi rõ nguồn trong bài làm. Bạn không được tham khảo bài làm của các bạn năm trước (vì nếu làm vậy thì bao giờ bạn mới có thể tự mình suy nghĩ để giải quyết vấn đề); sau khi kết thúc môn học, bạn cũng không được đưa bài làm cho các bạn khóa sau hoặc public bài làm trên Github (vì nếu làm vậy thì sẽ ảnh hưởng tới việc học của các bạn khóa sau). Nếu bạn có thể làm theo những gì mình nói thì điểm của bạn có thể sẽ không cao nhưng bạn sẽ có được những bước tiến thật sự. <font color=red>Trong trường hợp bạn vi phạm những điều mình nói ở trên thì sẽ bị 0 điểm cho toàn bộ môn học.</font>*

**Cách nộp bài**

Khi chấm bài, đầu tiên mình sẽ chọn `Kernel` - `Restart Kernel & Run All Cells`, để restart và chạy tất cả các cell trong notebook của bạn; do đó, trước khi nộp bài, bạn nên chạy thử `Kernel` - `Restart Kernel & Run All Cells` để đảm bảo mọi chuyện diễn ra đúng như mong đợi.

Sau đó, bạn tạo thư mục nộp bài theo cấu trúc sau:
- Thư mục `MSSV` (vd, nếu bạn có MSSV là 1234567 thì bạn đặt tên thư mục là `1234567`)
    - Thư mục `Lab01_ThuThapDuLieu`
        - File `Lab01_ThuThapDuLieu.ipynb` (không cần nộp các file khác)

Cuối cùng, bạn nén thư mục `MSSV` này lại và nộp ở link trên moodle. **Đuôi của file nén phải là .zip (chứ không được .rar hay gì khác).**

<font color=red>Bạn lưu ý tuân thủ chính xác qui định nộp bài ở trên.</font>

## Môi trường code

Ta thống nhất trong môn này: dùng phiên bản các package như trong file "min_ds-env.yml" 

Cách tạo ra môi trường code từ file "min_ds-env.yml" đã được nói ở file "02_BeforeClass-Notebook_Python.pdf".

Check môi trường code:

In [1]:
import sys
sys.executable

'/home/lekiet/miniconda3/envs/min_ds-env/bin/python'

Nếu không có vấn đề gì thì file chạy python sẽ là file của môi trường code "min_ds-env".

***
## Import

In [4]:
import requests
import requests_cache
from bs4 import BeautifulSoup
import time
import json
import re
import pandas as pd # Dùng để đọc và hiển thị file csv/tsv
from datetime import datetime, timedelta # Dùng để xử lý dữ liệu thời gian
# YOUR CODE HERE (OPTION) 
# Nếu cần các thư viện khác thì bạn có thể import ở đây

---

## Setup ở mức toàn cục (1đ)

Trong bài này, sẽ có hai bài con: một bài là thu thập dữ liệu từ web bằng cách parse HTML, một bài là dùng web API. 
- Khi parse HTML, để tránh "hit" trang web nhiều lần trong một khoảng thời gian ngắn thì bạn sẽ cho chương trình sleep giữa các lần thực hiện request. 
- Khi dùng web API thì không cần cho sleep giữa các lần thực hiện request nhưng sẽ cần sleep ở một chỗ khác (ở chỗ nào thì sẽ rõ khi đến phần dùng web API). 

Tóm lại là: trong các phần code mà bạn sắp sửa làm thì sẽ cần cho chương trình sleep. Như vậy, đầu tiên, bạn sẽ setup biến `sleep_time` toàn cục (nói là toàn cục nhưng thật ra chỉ cần biến bình thường thôi) để qui định số giây mà chương trình sẽ sleep (trong bài này cho 1 giây). Trong các hàm thì bạn có thể truy xuất được biến `sleep_time` này. Vì phần này rất khó nên mình sẽ làm luôn cho bạn ;-).

In [5]:
sleep_time = 1 # Đơn vị: giây

Xong rồi đó ;-).

Ngoài biến `sleep_time` thì bạn cũng cần setup/install cache để giúp giảm số lần phải lấy dữ liệu từ trên mạng xuống. Bạn setup cache như sau:

- Cho thời gian hết hạn cache là không bao giờ hết hạn.
- Không cache những response (kết quả khi thực hiện `requests.get`) mà trong thuộc tính `text` có chứa chuỗi `"incomplete_results":true` (tại sao làm như vậy thì sẽ rõ khi đến phần dùng web API). 

Gợi ý: trong hàm `requests_cache.install_cache` có tham số `filter_fn`; bạn có thể truyền vào tham số này một hàm mà có input là đối tượng response và trả về True nếu bạn muốn cache lại và False nếu không muốn.

In [6]:
# YOUR CODE HERE
requests_cache.install_cache(
    cache_name='cache',
    filter_fn = lambda x: x.text.find('"incomplete_results":true') == -1 #nếu ko có 'incom...:true' -> cache lại
)

In [5]:
# TEST
url1 = 'https://raw.githubusercontent.com/KienTrann/requests-cache-testing/main/should_be_cached.txt'
r = requests.get(url1) # Thực hiện request lần 1
r = requests.get(url1) # Thực hiện request lần 2
print(f'Text from url1:\n{r.text}')
assert r.from_cache == True

url2 = 'https://raw.githubusercontent.com/KienTrann/requests-cache-testing/main/should_not_be_cached.txt'
r = requests.get(url2) # Thực hiện request lần 1
r = requests.get(url2) # Thực hiện request lần 2
print('---')
print(f'Text from url2:\n{r.text}')
assert r.from_cache == False


Text from url1:
abc
xyz
"incomplete_results":false

---
Text from url2:
abc
xyz
"incomplete_results":true



---

## Thu thập dữ liệu từ web bằng cách parse HTML (4đ)

[Coursera](https://coursera.org) là trang cung cấp các khóa học online, là một nguồn tài liệu hữu ích để học tập. Trong phần này, bạn sẽ thu thập dữ liệu về các khóa học ở Coursera về một chủ đề nào đó. Mình có check ["Terms of Use" của Coursera](https://www.coursera.org/about/terms) thì không thấy nói rõ là cấm parse HTML (bạn check lại thử xem); với mục đích học thì chắc là không có vấn đề gì lớn, miễn là chương trình của bạn đừng "hit" trang web quá nhiều lần trong một khoảng thời gian ngắn. 

Công việc cụ thể của bạn trong phần này là viết hàm `collect_courses` ở bên dưới. Hàm này có các input như sau:
- `course_urls_file`: là chuỗi, cho biết tên của file txt chứa các đường link đến tất cả khóa học của Coursera, mỗi đường link nằm trên một dòng; bạn sẽ đi theo đường link để lấy thông tin của khóa học. File này đã được cung cấp cho bạn: file "course_urls.txt" (mình tạo ra file này bằng cách parse [trang XML này](https://www.coursera.org/sitemap~www~courses.xml)).
- `key_words`: là list, cho biết các từ khóa để lọc ra các link cần lấy. Ví dụ: `key_words=['data-science', 'datascience']` sẽ chỉ thu thập thông tin từ các link mà có từ "data-science" HOẶC "datascience" (có ở tên đường link, chẳng hạn: https://www.coursera.org/learn/spatial-data-science).
- `courses_file`: là chuỗi, cho biết tên của file tsv dùng để lưu dữ liệu thu thập được; đây là file output (file này không có sẵn mà sẽ được tạo ra sau khi chạy hàm của bạn). Mình có cung cấp file output đúng "correct_courses.tsv" (khi gọi hàm `collect_courses` với `key_words=['data-science', 'datascience']`) để bạn hình dung được nội dung cần có của file output. File này có các phần tử trên mỗi dòng được phân tách nhau bởi tab (`\t`), và file này mã hóa utf-8. File này gồm có các cột:
    - `url`: đường link đến khóa học.
    - `title`: tựa đề của khóa học.
    - `provider`: trường cung cấp khóa học.
    - `rating`: điểm số của khóa học (tại thời điểm bạn thu thập dữ liệu thì giá trị của cột này có thể sẽ hơi khác so với file output đúng của mình, trong trường hợp chưa có ai rate khóa học thì bạn có thể cho `rating` bằng `None`).
    - `num_ratings`: số lượng người đã cho điểm cho khóa học (tại thời điểm bạn thu thập dữ liệu thì giá trị của cột này có thể sẽ hơi khác so với file output đúng của mình, trong trường hợp chưa có ai rate khóa học thì bạn có thể cho `num_rating` bằng `None`).

Trong hàm này, khi thực hiện một request, nếu kết quả không được lấy từ cache thì bạn cần cho chương trình sleep rồi mới tiếp tục.

In [6]:
# Xem 5 dòng đầu của file "correct_courses.tsv"
df = pd.read_csv('correct_courses.tsv', sep='\t')
df.head()

url  \
0  https://www.coursera.org/learn/advanced-clinic...   
1  https://www.coursera.org/learn/advanced-data-s...   
2  https://www.coursera.org/learn/applied-data-sc...   
3  https://www.coursera.org/learn/applied-data-sc...   
4  https://www.coursera.org/learn/applied-data-sc...   

                                    title                       provider  \
0          Advanced Clinical Data Science  University of Colorado System   
1          Advanced Data Science Capstone                            IBM   
2           Applied Data Science Capstone                            IBM   
3         كابستون علوم البيانات التطبيقية                            IBM   
4  Applied Data Science for Data Analysts                     Databricks   

  rating num_ratings  
0   None        None  
1    4.6        360   
2    4.7      5,549   
3   None        None  
4    4.2         24

In [8]:
def collect_courses(course_urls_file, key_words, courses_file):
    
    # YOUR CODE HERE
    with open(course_urls_file) as f:
        filterd_courses_url = []
        
        for line in f:
            if(line != ' ' and any(kw in line for kw in key_words)):
                filterd_courses_url.append(line[:-1]) #take the string without '\n' at the end of the string 

    results = []
    for url in filterd_courses_url:
        time.sleep(sleep_time)
        res = [url]
        r = requests.get(url)
        soup = BeautifulSoup(r.content, 'html.parser')
        
        title = soup.find('h1', attrs={'data-e2e':'xdp-banner-title'})
        title = title.text
        res.append(title)

        provider = soup.find_all('h3', attrs= {'class':'headline-4-text bold rc-Partner__title'})
        provider = [x.text for x in provider] #take the .text of each tag found
        provider = ' '.join(provider) #join all of them into a string 
        res.append(provider) 

        rating = soup.find('span', attrs= {'data-test':'number-star-rating'})
        rating = 'None' if rating == None else rating.text.replace('stars', '')
        res.append(rating)   

        num_rating = soup.find('span', attrs={'data-test':'ratings-count-without-asterisks'})
        num_rating = 'None' if num_rating == None else num_rating.text.split(' ')[0]
        res.append(num_rating)

        results.append(res)

    #write to file 'output.txt'    
    dframe = pd.DataFrame(results)
    dframe.to_csv(courses_file, sep= '\t', header= ['url','title','provider','rating','num_ratings'], index= False)

In [10]:
# TEST
collect_courses('course_urls.txt', ['data-science', 'datascience'], 'courses.tsv')
courses = pd.read_csv('courses.tsv', sep='\t')

# Test số dòng và số cột
assert courses.shape == (46, 5) 

# Test số lượng giá trị thiếu của rating
assert (courses.rating == 'None').sum() >= 15 


In [ ]:
# TEST (bằng mắt)
courses

***
## Thu thập dữ liệu từ web bằng cách dùng web API

### Thu thập ít hơn hoặc bằng 1000 thùng chứa (2.5đ)

Đầu tiên, bạn sẽ viết hàm `collect_repositories` ở bên dưới. Hàm này có các input:
- `start_date` và `end_date`: là chuỗi, cho biết là muốn tìm các thùng chứa được tạo từ ngày nào đến ngày nào (bao gồm cả 2 đầu). Ví dụ, `start_date='2017-01-01'` (năm-tháng-ngày) và `end_date='2017-01-31'`.
- `keyword`: là chuỗi, cho biết là muốn tìm kiếm các thùng chứa với từ khóa nào (xuất hiện ở tên thùng chứa hoặc phần mô tả của thùng chứa). Ví dụ, `keyword='data science'`.
- `per_page`: là số nguyên, cho biết là muốn bao nhiêu kết quả trên một page. Hiện giờ GitHub cho tối đa là `per_page=100`.

Trong hàm này bạn sẽ vào đường link `f'https://api.github.com/search/repositories?q={keyword} created:{start_date}..{end_date}&per_page={per_page}'` (dạng f-string trong Python) để thu thập dữ liệu. Ví dụ, bạn có thể vào thử ở web browser: https://api.github.com/search/repositories?q=data%20science%20created:2017-01-01..2017-01-31&per_page=100 (khi nhập một đường link vào web browser và enter thì một số ký tự sẽ bị mã hóa; ví dụ khoảng trắng sẽ bị mã hóa là %20 hoặc +), hoặc dùng thư viện Requests:

In [12]:
# Thư viện requests sẽ tự động mã hóa khoảng trắng cho bạn
# Bạn có thể xem đường link sau khi đã mã hóa bằng câu lệnh: r.url
r = requests.get('https://api.github.com/search/repositories?q=data science created:2017-01-01..2017-01-31&per_page=100')

In [13]:
# Xem 100 ký tự đầu tiên trong chuỗi JSON lấy được
print(r.text[:100])

{"total_count":1546,"incomplete_results":false,"items":[{"id":79030157,"node_id":"MDEwOlJlcG9zaXRvcn


Bạn thấy "total_count" có giá trị là 1546, nghĩa là tìm được 1546 thùng chứa (tại thời điểm bạn chạy thì có thể khác một ít, vì có thể xảy ra trường hợp chủ thùng chứa xóa thùng chứa đi, hoặc chuyển sang private). Tuy nhiên, đây chỉ là trang đầu tiên và chỉ có 100 thùng chứa (số lượng phần tử của "items"). Link tới trang kế tiếp nằm trong headers của nội dung gửi về từ server.

In [13]:
r.headers['Link']

'<https://api.github.com/search/repositories?q=data+science+created%3A2017-01-01..2017-01-31&per_page=100&page=2>; rel="next", <https://api.github.com/search/repositories?q=data+science+created%3A2017-01-01..2017-01-31&per_page=100&page=10>; rel="last"'

Link tới trang kế tiếp là link mà ở bên cạnh phía phải có từ `rel="next"`. `rel="last"` nghĩa là trang cuối. Vì đây là trang đầu tiên nên chỉ có "next" và "last", trang ở giữa còn có thêm "first" (trang đầu) và "prev" (trang trước). Bạn có thể xem thêm ở [document](https://docs.github.com/en/free-pro-team@latest/rest/guides/traversing-with-pagination).

Nếu để ý thì bạn thấy trang cuối là `page=10`, nghĩa là tất cả các trang chỉ có $10\times100=1000$ thùng chứa. Trong khi đó, "total_count" là 1546. Lý do: "the GitHub Search API provides up to **1,000 results for each search**" ([document](https://docs.github.com/en/free-pro-team@latest/rest/reference/search)). Trong hàm này chỉ yêu cầu bạn lấy hết kết quả ở tất cả các trang là được ($\le1000$ kết quả). Việc lấy $>1000$ kết quả sẽ làm ở phần kế tiếp.

Một yêu cầu nữa trong hàm này là: 
- Khi request không được (`r.ok` có giá trị là `False`) do hết số lượng request được phép trong 1 phút (với API không chứng thực thì hiện nay được 10 request / phút) thì bạn sẽ cho chương trình sleep rồi dậy thực hiện request lại, nếu vẫn chưa được thì lại sleep ... Bạn lưu ý là, <font color=red>không được giả định lần request đầu tiên là chắc chắn thành công</font> (vì khi chấm bài mình sẽ phải chạy nhiều bài và có thể đến bài của bạn đã hết lượt request được phép).
- Khi request được nhưng giá trị của `incomplete_results`(xem thêm ở [document](https://docs.github.com/en/free-pro-team@latest/rest/reference/search#timeouts-and-incomplete-results)) bằng `True` thì bạn cần thực hiện request lại cho đến khi `incomplete_results` bằng `False`. Bạn còn nhớ khi setup cache ở đầu, mình yêu cầu bạn không cache lại các response mà phần text có chứa chuỗi `"incomplete_results":True`. Vì nếu bạn cache lại các response này thì ở lần request kế sẽ được lấy từ cache và `incomplete_results` luôn bằng `True`.

Output trả về của hàm `collect_repositories`:
- `repositories`: là list các thùng chứa lấy được (list gồm "item" của tất cả các trang).

In [7]:
def collect_repositories(start_date, end_date, keyword='data science', per_page=100):
    repositories = []
    # YOUR CODE HERE
    #ex: https://api.github.com/search/repositories?q=data%20science%20created:2017-01-01..2017-01-15&per_page=100
    url = f'https://api.github.com/search/repositories?q={keyword} created:{start_date}..{end_date}&per_page={per_page}'
    page = 0

    while(1):   
        #preconditions: check if the page is 'valid' to continue
        while(1):
            r = requests.get(url)
            response = json.loads(r.content)

            if r.ok == False or response['incomplete_results'] == True:
                time.sleep(sleep_time)
            else:
                break
        
        # process each item (= each dict) in items 
        items = response['items']
        count_item = 0
        
        for item in items:
            if count_item <= per_page:   
                repositories.append(item)
                count_item += 1
            else:
                break
        
        # go to next page
        # example: '<https://api.github.com/search/repositories?q=data+science+created%3A2017-01-01..2017-01-31&per_page=100&page=2>; rel="next", <https://api.github.com/search/repositories?q=data+science+created%3A2017-01-01..2017-01-31&per_page=100&page=10>; rel="last"'
        if per_page >= response['total_count']: # this is the first & also the last page
            break
        else:
            list_links = r.headers['Link']
            match = re.search('rel="next"', list_links)
            
            if not match: 
                break
            else:
                end_boundary = match.start()-3
                if page == 0: start_boundary = 1
                else: start_boundary = re.search('rel="prev"', list_links).end() + 3
                url = list_links[start_boundary:end_boundary]
            page += 1
        
    return repositories 

In [10]:
# DRAFT
start_date = '2017-01-02'
end_date = '2017-01-02'
keyword = 'data science'
per_page = 44
repositories = collect_repositories(start_date, end_date, keyword, per_page)
print(len(repositories))

44


In [7]:
# TEST
start_date = '2017-01-01'
end_date = '2017-01-15'
keyword = 'data science'
per_page = 100
repositories = collect_repositories(start_date, end_date, keyword, per_page)
print(len(repositories))

url = f'https://api.github.com/search/repositories?q={keyword} created:{start_date}..{end_date}'
while True:
    json_str = requests.get(url).text
    if '"incomplete_results":false' in json_str:
        i = json_str.find(':')
        j = json_str.find(',')
        total_count = int(json_str[i+1:j])
        break
     
assert len(repositories) == total_count 


663


### Thu thập nhiều hơn 1000 thùng chứa (2.5đ)

Để thu thập hết 1546 thùng chứa trong ví dụ ở trên, ta sẽ: 
- Chia khoảng thời gian các thùng chứa được tạo mà ta cần tìm kiếm ra các khoảng thời gian nhỏ hơn. Ví dụ chia 2017-01-01..2017-01-15 thành 2017-01-01..2017-01-07, 2017-01-08..2017-01-14, 2017-01-15..2017-01-15; ở đây, mỗi khoảng thời gian nhỏ gồm 7 ngày (kể cả 2 đầu), khoảng thời gian nhỏ sau cùng có thể sẽ không đủ 7 ngày (do bị lẻ). Trong Python, để làm việc với dữ liệu thời gian, bạn có thể dùng thư viện `datetime`; bạn có thể tự tìm hiểu về thư viện này ở: [document](https://docs.python.org/3.7/library/datetime.html), [Corey's video](https://www.youtube.com/watch?v=eirjjyP2qcQ), Google.
- Tìm kiếm với khoảng thời gian nhỏ hơn này bằng hàm `collect_repositories` ở trên (với khoảng thời gian đủ nhỏ sẽ có $\le1000$ thùng chứa).
- Kết hợp kết quả từ các lần tìm kiếm với các khoảng thời gian nhỏ này lại.

Bạn sẽ cài đặt các bước này ở hàm `collect_all_repositories` dưới đây. Hàm này có các input:
- `start_date`, `end_date`, `keyword`, `per_page`: các input này giống như ở hàm `collect_repositories`.
- `step`: là số nguyên, cho biết khoảng thời gian nhỏ gồm bao nhiêu ngày (tính cả 2 đầu). Trong ví dụ ở trên, `step=7`.

Output trả về của hàm này:
- `all_repositories`: là list tất cả các thùng chứa ("item") lấy được.

In [11]:
def collect_all_repositories(start_date, end_date, step, keyword='data science', per_page=100):
    all_repositories = []
    # YOUR CODE HERE
    # url = f'https://api.github.com/search/repositories?q={keyword} created:{start_date}..{end_date}'
    start_date, end_date = datetime.strptime(start_date, '%Y-%m-%d'), datetime.strptime(end_date, '%Y-%m-%d') # datetime objs
    tdelta = timedelta(days= step - 1)
    flag = 0

    while(flag == 0):
        start, end = start_date, start_date + tdelta 
        if end >= end_date:
            end = end_date
            flag = 1
        
        repo = collect_repositories(start.strftime('%Y-%m-%d'), end.strftime('%Y-%m-%d'), keyword, per_page)
        all_repositories += repo # list of lists
        start_date = end + timedelta(days= 1)

    return all_repositories

In [ ]:
# DRAFT
start_date = '2017-01-01'
end_date = '2017-01-31'
step = 10
keyword = 'data science'
per_page = 200
all_repositories = collect_all_repositories(start_date, end_date, step, keyword, per_page)
print(len(all_repositories))

In [12]:
# TEST
start_date = '2017-01-01'
end_date = '2017-01-31'
step = 10
keyword = 'data science'
per_page = 100
all_repositories = collect_all_repositories(start_date, end_date, step, keyword, per_page)
print(len(all_repositories))

url = f'https://api.github.com/search/repositories?q={keyword} created:{start_date}..{end_date}'
while True:
    json_str = requests.get(url).text
    if '"incomplete_results":false' in json_str:
        i = json_str.find(':')
        j = json_str.find(',')
        total_count = int(json_str[i+1:j])
        break

assert len(all_repositories) == total_count 


1546
